// @openai/guardrails/dist/client.js
"use strict";
/**
 * High-level GuardrailsClient for easy integration with OpenAI APIs.
 *
 * This module provides GuardrailsOpenAI and GuardrailsAzureOpenAI classes that
 * subclass OpenAI's clients to provide full API compatibility while automatically
 * applying guardrails to text-based methods that could benefit from validation.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GuardrailsAzureOpenAI = exports.GuardrailsOpenAI = void 0;
const openai_1 = require("openai");
const base_client_1 = require("./base-client");
/**
 * OpenAI subclass with automatic guardrail integration.
 *
 * This class provides full OpenAI API compatibility while automatically
 * applying guardrails to text-based methods that could benefit from validation.
 *
 * Methods with guardrails:
 * - chat.completions.create() - Input/output validation
 * - responses.create() - Input/output validation
 *
 * All other methods pass through unchanged for full API compatibility.
 */
class GuardrailsOpenAI extends openai_1.OpenAI {
    constructor(guardrailsClient, options) {
        // Initialize OpenAI client first
        super(options);
        // Store the initialized guardrails client
        this.guardrailsClient = guardrailsClient;
        // Override chat and responses after initialization
        this.overrideResources();
    }
    /**
     * Create a new GuardrailsOpenAI instance.
     *
     * @param config Pipeline configuration (file path, object, or JSON string)
     * @param options Optional OpenAI client options
     * @param raiseGuardrailErrors If true, raise exceptions when guardrails fail to execute.
     *   If false (default), treat guardrail execution errors as safe and continue.
     *   Note: Tripwires (guardrail violations) are handled separately and not affected
     *   by this parameter.
     * @returns Promise resolving to configured GuardrailsOpenAI instance
     */
    static async create(config, options, raiseGuardrailErrors = false) {
        // Create and initialize the guardrails client
        const guardrailsClient = new GuardrailsBaseClientImpl();
        await guardrailsClient.initializeClient(config, options || {}, openai_1.OpenAI);
        // Store the raiseGuardrailErrors setting
        guardrailsClient.raiseGuardrailErrors = raiseGuardrailErrors;
        // Create the instance with the initialized client
        return new GuardrailsOpenAI(guardrailsClient, options);
    }
    /**
     * Override chat and responses with our guardrail-enhanced versions.
     */
    overrideResources() {
        const { Chat } = require('./resources/chat');
        const { Responses } = require('./resources/responses');
        // Replace the chat and responses attributes with our versions
        Object.defineProperty(this, 'chat', {
            value: new Chat(this.guardrailsClient),
            writable: false,
            configurable: false,
        });
        Object.defineProperty(this, 'responses', {
            value: new Responses(this.guardrailsClient),
            writable: false,
            configurable: false,
        });
        Object.defineProperty(this, 'guardrails', {
            value: {
                responses: new Responses(this.guardrailsClient),
                chat: new Chat(this.guardrailsClient),
            },
            writable: false,
            configurable: false,
        });
    }
}
exports.GuardrailsOpenAI = GuardrailsOpenAI;
// ---------------- Azure OpenAI Variant -----------------
/**
 * Azure OpenAI subclass with automatic guardrail integration.
 */
class GuardrailsAzureOpenAI extends openai_1.AzureOpenAI {
    constructor(guardrailsClient, azureArgs) {
        // Initialize Azure OpenAI client first
        super(azureArgs);
        // Store the initialized guardrails client
        this.guardrailsClient = guardrailsClient;
        // Override chat and responses after initialization
        this.overrideResources();
    }
    /**
     * Create a new GuardrailsAzureOpenAI instance.
     *
     * @param config Pipeline configuration (file path, object, or JSON string)
     * @param azureOptions Azure OpenAI client options
     * @param raiseGuardrailErrors If true, raise exceptions when guardrails fail to execute.
     *   If false (default), treat guardrail execution errors as safe and continue.
     *   Note: Tripwires (guardrail violations) are handled separately and not affected
     *   by this parameter.
     * @returns Promise resolving to configured GuardrailsAzureOpenAI instance
     */
    static async create(config, azureOptions, raiseGuardrailErrors = false) {
        // Create and initialize the guardrails client
        const guardrailsClient = new GuardrailsBaseClientImplAzure();
        await guardrailsClient.initializeClient(config, azureOptions, openai_1.AzureOpenAI);
        // Store the raiseGuardrailErrors setting
        guardrailsClient.raiseGuardrailErrors = raiseGuardrailErrors;
        // Create the instance with the initialized client
        return new GuardrailsAzureOpenAI(guardrailsClient, azureOptions);
    }
    /**
     * Override chat and responses with our guardrail-enhanced versions.
     */
    overrideResources() {
        const { Chat } = require('./resources/chat');
        const { Responses } = require('./resources/responses');
        // Replace the chat and responses attributes with our versions
        Object.defineProperty(this, 'chat', {
            value: new Chat(this.guardrailsClient),
            writable: false,
            configurable: false,
        });
        Object.defineProperty(this, 'responses', {
            value: new Responses(this.guardrailsClient),
            writable: false,
            configurable: false,
        });
        Object.defineProperty(this, 'guardrails', {
            value: {
                responses: new Responses(this.guardrailsClient),
                chat: new Chat(this.guardrailsClient),
            },
            writable: false,
            configurable: false,
        });
    }
}
exports.GuardrailsAzureOpenAI = GuardrailsAzureOpenAI;
/**
 * Concrete implementation of GuardrailsBaseClient.
 */
class GuardrailsBaseClientImpl extends base_client_1.GuardrailsBaseClient {
    /**
     * Create default context with guardrail_llm client.
     */
    createDefaultContext() {
        // Create a separate client instance for guardrails (not the same as main client)
        const guardrailClient = new openai_1.OpenAI({
            apiKey: this._resourceClient.apiKey,
            baseURL: this._resourceClient.baseURL,
            organization: this._resourceClient.organization,
            timeout: this._resourceClient.timeout,
            maxRetries: this._resourceClient.maxRetries,
        });
        return {
            guardrailLlm: guardrailClient,
        };
    }
    /**
     * Override resources with guardrail-enhanced versions.
     * Not used in the concrete implementation since the main classes handle this.
     */
    overrideResources() {
        // No-op in the implementation class
    }
}
/**
 * Azure-specific implementation of GuardrailsBaseClient.
 */
class GuardrailsBaseClientImplAzure extends base_client_1.GuardrailsBaseClient {
    constructor() {
        super(...arguments);
        this.azureArgs = {};
    }
    /**
     * Create default context with Azure guardrail_llm client.
     */
    createDefaultContext() {
        // Create a separate Azure client instance for guardrails
        const guardrailClient = new openai_1.AzureOpenAI(this.azureArgs);
        return {
            guardrailLlm: guardrailClient,
        };
    }
    /**
     * Override resources with guardrail-enhanced versions.
     * Not used in the concrete implementation since the main classes handle this.
     */
    overrideResources() {
        // No-op in the implementation class
    }
    /**
     * Store Azure args for creating guardrail client.
     */
    async initializeClient(config, openaiArgs, clientClass) {
        // Store azure arguments
        this.azureArgs = openaiArgs;
        // Call parent initialization
        return super.initializeClient(config, openaiArgs, clientClass);
    }
}
