// @openai/guardrails/dist/exceptions.js
"use strict";
/**
 * Exception types for Guardrails.
 *
 * This module provides custom error classes for guardrail-related errors.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GuardrailExecutionError = exports.GuardrailNotFoundError = exports.GuardrailConfigurationError = exports.GuardrailTripwireTriggered = exports.GuardrailError = void 0;
/**
 * Base class for all guardrail-related errors.
 */
class GuardrailError extends Error {
    constructor(message) {
        super(message);
        this.name = 'GuardrailError';
    }
}
exports.GuardrailError = GuardrailError;
/**
 * Exception raised when a guardrail tripwire is triggered.
 *
 * This exception indicates that a guardrail check has identified
 * a critical failure that should prevent further processing.
 */
class GuardrailTripwireTriggered extends GuardrailError {
    constructor(guardrailResult) {
        const message = `Guardrail tripwire triggered: ${guardrailResult.info?.guardrail_name || 'Unknown'}`;
        super(message);
        this.name = 'GuardrailTripwireTriggered';
        this.guardrailResult = guardrailResult;
    }
}
exports.GuardrailTripwireTriggered = GuardrailTripwireTriggered;
/**
 * Exception raised when there's an issue with guardrail configuration.
 */
class GuardrailConfigurationError extends GuardrailError {
    constructor(message) {
        super(message);
        this.name = 'GuardrailConfigurationError';
    }
}
exports.GuardrailConfigurationError = GuardrailConfigurationError;
/**
 * Exception raised when a guardrail check function is not found.
 */
class GuardrailNotFoundError extends GuardrailError {
    constructor(name) {
        super(`Guardrail '${name}' not found`);
        this.name = 'GuardrailNotFoundError';
    }
}
exports.GuardrailNotFoundError = GuardrailNotFoundError;
/**
 * Exception raised when there's an issue with guardrail execution.
 */
class GuardrailExecutionError extends GuardrailError {
    constructor(message, cause) {
        super(message);
        this.name = 'GuardrailExecutionError';
        this.cause = cause;
    }
}
exports.GuardrailExecutionError = GuardrailExecutionError;
