// @openai/guardrails/dist/index.js
"use strict";
/**
 * Guardrails public API surface.
 *
 * This package exposes utilities to define and run guardrails which validate
 * arbitrary data. The submodules provide runtime helpers, exception
 * types and a registry of built-in checks.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.cli = exports.GuardrailAgent = exports.GuardrailsBaseClient = exports.GuardrailsAzureOpenAI = exports.GuardrailsOpenAI = exports.loadPipelineBundles = exports.loadConfigBundleFromFile = exports.loadConfigBundle = exports.instantiateGuardrails = exports.runGuardrails = exports.checkPlainText = exports.ConfiguredGuardrail = exports.GuardrailSpec = exports.defaultSpecRegistry = exports.GuardrailRegistry = exports.GuardrailExecutionError = exports.GuardrailNotFoundError = exports.GuardrailConfigurationError = exports.GuardrailTripwireTriggered = exports.GuardrailError = void 0;
// Exception types
var exceptions_1 = require("./exceptions");
Object.defineProperty(exports, "GuardrailError", { enumerable: true, get: function () { return exceptions_1.GuardrailError; } });
Object.defineProperty(exports, "GuardrailTripwireTriggered", { enumerable: true, get: function () { return exceptions_1.GuardrailTripwireTriggered; } });
Object.defineProperty(exports, "GuardrailConfigurationError", { enumerable: true, get: function () { return exceptions_1.GuardrailConfigurationError; } });
Object.defineProperty(exports, "GuardrailNotFoundError", { enumerable: true, get: function () { return exceptions_1.GuardrailNotFoundError; } });
Object.defineProperty(exports, "GuardrailExecutionError", { enumerable: true, get: function () { return exceptions_1.GuardrailExecutionError; } });
// Registry and specifications
var registry_1 = require("./registry");
Object.defineProperty(exports, "GuardrailRegistry", { enumerable: true, get: function () { return registry_1.GuardrailRegistry; } });
Object.defineProperty(exports, "defaultSpecRegistry", { enumerable: true, get: function () { return registry_1.defaultSpecRegistry; } });
var spec_1 = require("./spec");
Object.defineProperty(exports, "GuardrailSpec", { enumerable: true, get: function () { return spec_1.GuardrailSpec; } });
// Runtime execution
var runtime_1 = require("./runtime");
Object.defineProperty(exports, "ConfiguredGuardrail", { enumerable: true, get: function () { return runtime_1.ConfiguredGuardrail; } });
Object.defineProperty(exports, "checkPlainText", { enumerable: true, get: function () { return runtime_1.checkPlainText; } });
Object.defineProperty(exports, "runGuardrails", { enumerable: true, get: function () { return runtime_1.runGuardrails; } });
Object.defineProperty(exports, "instantiateGuardrails", { enumerable: true, get: function () { return runtime_1.instantiateGuardrails; } });
Object.defineProperty(exports, "loadConfigBundle", { enumerable: true, get: function () { return runtime_1.loadConfigBundle; } });
Object.defineProperty(exports, "loadConfigBundleFromFile", { enumerable: true, get: function () { return runtime_1.loadConfigBundleFromFile; } });
Object.defineProperty(exports, "loadPipelineBundles", { enumerable: true, get: function () { return runtime_1.loadPipelineBundles; } });
// Client interfaces (Drop-in replacements for OpenAI clients)
var client_1 = require("./client");
Object.defineProperty(exports, "GuardrailsOpenAI", { enumerable: true, get: function () { return client_1.GuardrailsOpenAI; } });
Object.defineProperty(exports, "GuardrailsAzureOpenAI", { enumerable: true, get: function () { return client_1.GuardrailsAzureOpenAI; } });
// Base client functionality
var base_client_1 = require("./base-client");
Object.defineProperty(exports, "GuardrailsBaseClient", { enumerable: true, get: function () { return base_client_1.GuardrailsBaseClient; } });
// Built-in checks
// Importing this module will automatically register all built-in guardrails
// with the defaultSpecRegistry
__exportStar(require("./checks"), exports);
// Utility functions
__exportStar(require("./utils"), exports);
// Evaluation framework
__exportStar(require("./evals"), exports);
// Agents SDK integration
var agents_1 = require("./agents");
Object.defineProperty(exports, "GuardrailAgent", { enumerable: true, get: function () { return agents_1.GuardrailAgent; } });
// CLI tool
var cli_1 = require("./cli");
Object.defineProperty(exports, "cli", { enumerable: true, get: function () { return cli_1.main; } });
