// @openai/guardrails/dist/spec.js
"use strict";
/**
 * Guardrail specification and model resolution.
 *
 * This module defines the `GuardrailSpec` class, which captures the metadata,
 * configuration schema, and logic for a guardrail. It also includes a structured
 * metadata model for attaching descriptive and extensible information to guardrails,
 * and instantiation logic for producing executable guardrail instances.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.GuardrailSpec = void 0;
const runtime_1 = require("./runtime");
/**
 * Immutable descriptor for a registered guardrail.
 *
 * Encapsulates all static information about a guardrail, including its name,
 * human description, supported media type, configuration schema, the validation
 * function, context requirements, and optional metadata.
 *
 * GuardrailSpec instances are registered for cataloguing and introspection,
 * but should be instantiated with user configuration to create a runnable guardrail
 * for actual use.
 *
 * @warning The mediaType field determines which content types this guardrail can process.
 * Only guardrails with compatible media types will be executed. Use 'text/plain' for text content.
 */
class GuardrailSpec {
    constructor(name, description, mediaType, configSchema, checkFn, ctxRequirements, metadata) {
        this.name = name;
        this.description = description;
        this.mediaType = mediaType;
        this.configSchema = configSchema;
        this.checkFn = checkFn;
        this.ctxRequirements = ctxRequirements;
        this.metadata = metadata;
    }
    /**
     * Return the JSON schema for the guardrail's configuration model.
     *
     * This method provides the schema needed for UI validation, documentation,
     * or API introspection.
     *
     * @returns JSON schema describing the config model fields.
     */
    schema() {
        return this.configSchema._def;
    }
    /**
     * Produce a configured, executable guardrail from this specification.
     *
     * This is the main entry point for creating guardrail instances that can
     * be run in a validation pipeline. The returned object is fully bound to
     * this definition and the provided configuration.
     *
     * @param config Validated configuration for this guardrail.
     * @returns Runnable guardrail instance.
     */
    instantiate(config) {
        return new runtime_1.ConfiguredGuardrail(this, config);
    }
}
exports.GuardrailSpec = GuardrailSpec;
