// @openai/guardrails/dist/checks/competitors.js
"use strict";
/**
 * Competitor detection guardrail module.
 *
 * This module provides a guardrail for detecting mentions of competitors in text.
 * It uses case-insensitive keyword matching against a configurable list of competitor names.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.competitorsCheck = exports.CompetitorContext = exports.CompetitorConfig = void 0;
const zod_1 = require("zod");
const keywords_1 = require("./keywords");
const registry_1 = require("../registry");
/**
 * Configuration schema for competitor detection.
 *
 * This configuration is used to specify a list of competitor names that will be
 * flagged if detected in the analyzed text. Matching is case-insensitive.
 */
exports.CompetitorConfig = zod_1.z.object({
    /** List of competitor names to detect. Matching is case-insensitive. */
    keywords: zod_1.z.array(zod_1.z.string()).min(1),
});
/**
 * Context requirements for the competitors guardrail.
 */
exports.CompetitorContext = zod_1.z.object({});
/**
 * Guardrail function to flag competitor mentions in text.
 *
 * Checks the provided text for the presence of any competitor names specified
 * in the configuration. Returns a `GuardrailResult` indicating whether any
 * competitor keyword was found.
 *
 * @param ctx Context object for the guardrail runtime (unused).
 * @param data Text to analyze for competitor mentions.
 * @param config Configuration specifying competitor keywords.
 * @returns GuardrailResult indicating whether any competitor keyword was detected.
 */
const competitorsCheck = (ctx, data, config) => {
    // Convert to KeywordsConfig format and reuse the keywords check
    const keywordsConfig = {
        keywords: config.keywords,
    };
    const result = (0, keywords_1.keywordsCheck)(ctx, data, keywordsConfig);
    // Update the guardrail name in the result
    return {
        ...result,
        info: {
            ...result.info,
            guardrail_name: 'Competitors',
        },
    };
};
exports.competitorsCheck = competitorsCheck;
// Auto-register this guardrail with the default registry
registry_1.defaultSpecRegistry.register('Competitors', exports.competitorsCheck, 'Checks if the model output mentions any competitors from the provided list', 'text/plain', exports.CompetitorConfig, exports.CompetitorContext, { engine: 'regex' });
