// @openai/guardrails/dist/utils/context.js
"use strict";
/**
 * Utility helpers for dealing with guardrail execution contexts.
 *
 * This module exposes utilities to validate runtime objects against guardrail context schemas.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContextValidationError = void 0;
exports.validateGuardrailContext = validateGuardrailContext;
exports.hasProperty = hasProperty;
exports.hasRequiredProperties = hasRequiredProperties;
const exceptions_1 = require("../exceptions");
/**
 * Error thrown when context validation fails.
 */
class ContextValidationError extends exceptions_1.GuardrailError {
    constructor(message) {
        super(message);
        this.name = 'ContextValidationError';
    }
}
exports.ContextValidationError = ContextValidationError;
/**
 * Validates a context object against a guardrail's declared context schema.
 *
 * @param guardrail - Guardrail whose context requirements define the schema.
 * @param ctx - Application context instance to validate.
 * @throws {ContextValidationError} If ctx does not satisfy required fields.
 * @throws {TypeError} If ctx's attributes cannot be introspected.
 */
function validateGuardrailContext(guardrail, ctx) {
    const model = guardrail.definition.ctxRequirements;
    try {
        // For now, we'll do basic validation
        // In a full implementation, you might want to use a validation library like Zod or Joi
        if (model && typeof model === 'object') {
            // Check if required properties exist on the context
            for (const [key, value] of Object.entries(model)) {
                if (value && typeof value === 'object' && 'required' in value && value.required) {
                    if (!(key in ctx)) {
                        throw new ContextValidationError(`Context for '${guardrail.definition.name}' guardrail expects required property '${key}' which is missing from context`);
                    }
                }
            }
        }
    }
    catch (error) {
        if (error instanceof ContextValidationError) {
            throw error;
        }
        // Attempt to get application context schema for better error message
        let appCtxFields = {};
        try {
            appCtxFields = Object.getOwnPropertyNames(ctx).reduce((acc, prop) => {
                acc[prop] = typeof ctx[prop];
                return acc;
            }, {});
        }
        catch {
            const msg = `Context must support property access, please pass Context as a class instead of '${typeof ctx}'.`;
            throw new ContextValidationError(msg);
        }
        const ctxRequirements = model ? Object.keys(model) : [];
        const msg = `Context for '${guardrail.definition.name}' guardrail expects ${ctxRequirements} which does not match ctx schema '${Object.keys(appCtxFields)}': ${error}`;
        throw new ContextValidationError(msg);
    }
}
/**
 * Type guard to check if an object has a specific property.
 */
function hasProperty(obj, prop) {
    return prop in obj;
}
/**
 * Type guard to check if an object has all required properties.
 */
function hasRequiredProperties(obj, requiredProps) {
    return requiredProps.every((prop) => hasProperty(obj, prop));
}
