// @openai/guardrails/dist/utils/openai-vector-store.js
"use strict";
/**
 * OpenAI Vector Store Creation Utility
 *
 * This module provides utilities for creating OpenAI vector stores from files or directories,
 * providing functionality for creating OpenAI vector stores.
 *
 * Note: This implementation uses OpenAI v4 API.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createOpenAIVectorStoreFromPath = createOpenAIVectorStoreFromPath;
const openai_1 = __importDefault(require("openai"));
/**
 * Create an OpenAI vector store from files or directories.
 *
 * This function creates a vector store by:
 * 1. Creating an assistant with file search capabilities
 * 2. Uploading files to OpenAI
 * 3. Attaching files to the assistant
 *
 * @param path - Path to file or directory containing documents
 * @param config - Configuration for the OpenAI client
 * @returns Assistant ID that can be used as a knowledge source
 */
async function createOpenAIVectorStoreFromPath(path, config) {
    const client = new openai_1.default({ apiKey: config.apiKey });
    // Check if path exists
    try {
        const fs = await Promise.resolve().then(() => __importStar(require('fs/promises')));
        await fs.access(path);
    }
    catch {
        throw new Error(`Path does not exist: ${path}`);
    }
    try {
        // Get list of files to upload
        const filePaths = await getFilePaths(path);
        if (filePaths.length === 0) {
            throw new Error(`No supported files found in ${path}`);
        }
        // Upload files
        const fileIds = await uploadFiles(client, filePaths);
        if (fileIds.length === 0) {
            throw new Error('No files were successfully uploaded');
        }
        // Create a vector store
        const vectorStore = await client.vectorStores.create({
            name: config.name || `anti_hallucination_${path.split('/').pop() || 'documents'}`,
        });
        // Attach files to the vector store
        for (const fileId of fileIds) {
            await client.vectorStores.files.create(vectorStore.id, { file_id: fileId });
        }
        // Wait for files to be processed
        await waitForFileProcessing(client, fileIds);
        // Return the vector store ID
        return vectorStore.id;
    }
    catch (error) {
        throw new Error(`Failed to create vector store: ${error instanceof Error ? error.message : String(error)}`);
    }
}
/**
 * Get list of supported files from a path.
 */
async function getFilePaths(path) {
    const fs = await Promise.resolve().then(() => __importStar(require('fs/promises')));
    const pathModule = await Promise.resolve().then(() => __importStar(require('path')));
    const supportedFileTypes = [
        '.c',
        '.cpp',
        '.cs',
        '.css',
        '.doc',
        '.docx',
        '.go',
        '.html',
        '.java',
        '.js',
        '.json',
        '.md',
        '.pdf',
        '.php',
        '.pptx',
        '.py',
        '.rb',
        '.sh',
        '.tex',
        '.ts',
        '.txt',
    ];
    // Check extension before stat if it looks like a file
    const ext = pathModule.extname(path).toLowerCase();
    if (ext && !supportedFileTypes.includes(ext)) {
        // If the path has an extension and it's not supported, skip stat and return []
        return [];
    }
    try {
        const stat = await fs.stat(path);
        if (stat.isFile()) {
            // ext already calculated above
            return supportedFileTypes.includes(ext) ? [path] : [];
        }
        else if (stat.isDirectory()) {
            const files = [];
            const entries = await fs.readdir(path, { withFileTypes: true });
            for (const entry of entries) {
                if (entry.isFile()) {
                    const fullPath = pathModule.join(path, entry.name);
                    const entryExt = pathModule.extname(entry.name).toLowerCase();
                    if (supportedFileTypes.includes(entryExt)) {
                        files.push(fullPath);
                    }
                }
            }
            return files;
        }
    }
    catch (error) {
        throw new Error(`Error reading path ${path}: ${error instanceof Error ? error.message : String(error)}`);
    }
    return [];
}
/**
 * Upload files to OpenAI and return file IDs.
 */
async function uploadFiles(client, filePaths) {
    const fs = await Promise.resolve().then(() => __importStar(require('fs/promises')));
    const fileIds = [];
    for (const filePath of filePaths) {
        try {
            const fileBuffer = await fs.readFile(filePath);
            const pathModule = await Promise.resolve().then(() => __importStar(require('path')));
            const fileName = pathModule.basename(filePath);
            // Create a File-like object that matches the Uploadable interface
            const file = await client.files.create({
                file: new File([fileBuffer], fileName, { type: 'application/octet-stream' }),
                purpose: 'assistants',
            });
            fileIds.push(file.id);
        }
        catch (error) {
            console.warn(`Failed to upload file ${filePath}: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
    return fileIds;
}
/**
 * Wait for files to be processed by OpenAI.
 */
async function waitForFileProcessing(client, fileIds) {
    let completed = false;
    while (!completed) {
        const allCompleted = await Promise.all(fileIds.map(async (fileId) => {
            try {
                const file = await client.files.retrieve(fileId);
                return file.status === 'processed';
            }
            catch {
                return false;
            }
        }));
        if (allCompleted.every((status) => status)) {
            completed = true;
        }
        // Wait 1 second before checking again
        await new Promise((resolve) => setTimeout(resolve, 1000));
    }
}
