// @openai/guardrails/dist/utils/vector-store.js
"use strict";
/**
 * Utilities for creating and managing vector stores.
 *
 * This module provides utilities for working with embeddings and vector stores,
 * providing functionality for creating and managing vector stores.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createVectorStore = createVectorStore;
/**
 * Create a vector store based on configuration.
 *
 * @param config - Configuration for the vector store.
 * @returns A configured vector store instance.
 */
async function createVectorStore(config) {
    switch (config.type) {
        case 'memory':
            return new MemoryVectorStore(config.config);
        case 'pinecone':
            return new PineconeVectorStore(config.config);
        case 'weaviate':
            return new WeaviateVectorStore(config.config);
        case 'chroma':
            return new ChromaVectorStore(config.config);
        default:
            throw new Error(`Unsupported vector store type: ${config.type}`);
    }
}
/**
 * In-memory vector store implementation.
 */
class MemoryVectorStore {
    constructor(config) {
        this.config = config;
        this.documents = new Map();
        this.embeddings = new Map();
    }
    async addDocuments(documents) {
        for (const doc of documents) {
            this.documents.set(doc.id, doc);
            if (doc.embedding) {
                this.embeddings.set(doc.id, doc.embedding);
            }
        }
    }
    async search(query, limit = 10) {
        // Simple implementation - in a real scenario, you'd use proper similarity search
        const results = [];
        for (const [id, doc] of this.documents) {
            const embedding = this.embeddings.get(id);
            if (embedding) {
                // Simple cosine similarity (placeholder implementation)
                const score = this.cosineSimilarity([1, 0, 0], embedding); // Placeholder query embedding
                results.push({ document: doc, score });
            }
        }
        return results.sort((a, b) => b.score - a.score).slice(0, limit);
    }
    async deleteDocuments(documentIds) {
        for (const id of documentIds) {
            this.documents.delete(id);
            this.embeddings.delete(id);
        }
    }
    async getDocument(id) {
        return this.documents.get(id) || null;
    }
    cosineSimilarity(a, b) {
        if (a.length !== b.length)
            return 0;
        let dotProduct = 0;
        let normA = 0;
        let normB = 0;
        for (let i = 0; i < a.length; i++) {
            dotProduct += a[i] * b[i];
            normA += a[i] * a[i];
            normB += b[i] * b[i];
        }
        if (normA === 0 || normB === 0)
            return 0;
        return dotProduct / (Math.sqrt(normA) * Math.sqrt(normB));
    }
}
/**
 * Placeholder implementations for other vector store types.
 */
class PineconeVectorStore {
    constructor(config) {
        this.config = config;
    }
    async addDocuments(_documents) {
        throw new Error('Pinecone vector store not implemented');
    }
    async search(_query, _limit) {
        throw new Error('Pinecone vector store not implemented');
    }
    async deleteDocuments(_documentIds) {
        throw new Error('Pinecone vector store not implemented');
    }
    async getDocument(_id) {
        throw new Error('Pinecone vector store not implemented');
    }
}
class WeaviateVectorStore {
    constructor(config) {
        this.config = config;
    }
    async addDocuments(_documents) {
        throw new Error('Weaviate vector store not implemented');
    }
    async search(_query, _limit) {
        throw new Error('Weaviate vector store not implemented');
    }
    async deleteDocuments(_documentIds) {
        throw new Error('Weaviate vector store not implemented');
    }
    async getDocument(_id) {
        throw new Error('Weaviate vector store not implemented');
    }
}
class ChromaVectorStore {
    constructor(config) {
        this.config = config;
    }
    async addDocuments(_documents) {
        throw new Error('Chroma vector store not implemented');
    }
    async search(_query, _limit) {
        throw new Error('Chroma vector store not implemented');
    }
    async deleteDocuments(_documentIds) {
        throw new Error('Chroma vector store not implemented');
    }
    async getDocument(_id) {
        throw new Error('Chroma vector store not implemented');
    }
}
