// @openai/guardrails/dist/__tests__/integration/integration.test.js
"use strict";
/**
 * Integration tests for the guardrails system.
 *
 * This module tests the complete integration of all components including:
 * - Guardrail registration and execution
 * - Configuration bundle loading
 * - Error handling
 * - Performance and scalability
 */
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const registry_1 = require("../../registry");
const runtime_1 = require("../../runtime");
// Mock check function for testing
const mockCheck = (ctx, data) => ({
    tripwireTriggered: data === 'trigger',
    info: {
        sampled_text: data,
    },
});
(0, vitest_1.describe)('Integration Tests', () => {
    let registry;
    (0, vitest_1.beforeEach)(() => {
        registry = new registry_1.GuardrailRegistry();
        // Register test guardrails
        registry.register('test_guard', mockCheck, 'Test guardrail', 'text/plain');
        registry.register('trigger_guard', mockCheck, 'Trigger guardrail', 'text/plain');
    });
    (0, vitest_1.describe)('Guardrail Registration and Execution', () => {
        (0, vitest_1.it)('should register and execute guardrails', () => {
            const spec = registry.get('test_guard');
            (0, vitest_1.expect)(spec).toBeDefined();
            (0, vitest_1.expect)(spec.name).toBe('test_guard');
            const guardrail = spec.instantiate({});
            (0, vitest_1.expect)(guardrail).toBeDefined();
        });
        (0, vitest_1.it)('should handle multiple guardrails in sequence', () => {
            const spec1 = registry.get('test_guard');
            const spec2 = registry.get('trigger_guard');
            (0, vitest_1.expect)(spec1).toBeDefined();
            (0, vitest_1.expect)(spec2).toBeDefined();
            (0, vitest_1.expect)(spec1.name).toBe('test_guard');
            (0, vitest_1.expect)(spec2.name).toBe('trigger_guard');
        });
        (0, vitest_1.it)('should execute guardrails with different inputs', async () => {
            const spec = registry.get('test_guard');
            const guardrail = spec.instantiate({});
            // Test non-triggering input
            const result1 = await guardrail.run({}, 'safe data');
            (0, vitest_1.expect)(result1.tripwireTriggered).toBe(false);
            // Test triggering input
            const result2 = await guardrail.run({}, 'trigger');
            (0, vitest_1.expect)(result2.tripwireTriggered).toBe(true);
        });
    });
    (0, vitest_1.describe)('Configuration Bundle Loading', () => {
        (0, vitest_1.it)('should load and validate configuration bundle', () => {
            const bundleJson = JSON.stringify({
                version: 1,
                stageName: 'test',
                guardrails: [
                    {
                        name: 'test_guard',
                        config: { threshold: 10 },
                    },
                ],
            });
            const bundle = (0, runtime_1.loadConfigBundle)(bundleJson);
            (0, vitest_1.expect)(bundle.version).toBe(1);
            (0, vitest_1.expect)(bundle.stageName).toBe('test');
            (0, vitest_1.expect)(bundle.guardrails).toHaveLength(1);
        });
    });
    (0, vitest_1.describe)('Error Handling', () => {
        (0, vitest_1.it)('should handle invalid guardrail names gracefully', () => {
            const spec = registry.get('nonexistent_guard');
            (0, vitest_1.expect)(spec).toBeUndefined();
        });
        (0, vitest_1.it)('should handle malformed configuration bundles', () => {
            const invalidBundle = JSON.stringify({
                stageName: 'test',
                // Missing required fields
            });
            (0, vitest_1.expect)(() => (0, runtime_1.loadConfigBundle)(invalidBundle)).toThrow();
        });
        // TODO: Add test for runtime errors once registry mocking is resolved
        (0, vitest_1.it)('should have placeholder for runtime error tests', () => {
            (0, vitest_1.expect)(true).toBe(true);
        });
    });
    // TODO: Add performance tests once registry mocking is resolved
    (0, vitest_1.describe)('Performance and Scalability', () => {
        (0, vitest_1.it)('should have placeholder for performance tests', () => {
            (0, vitest_1.expect)(true).toBe(true);
        });
    });
});
