// @openai/guardrails/dist/__tests__/unit/evals.test.js
"use strict";
/**
 * Unit tests for the evaluation framework.
 *
 * This module tests the evaluation framework components including:
 * - GuardrailMetricsCalculator
 * - Dataset validation
 * - Results reporting
 * - Core evaluation functionality
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const evals_1 = require("../../evals");
// Using type assertion for fs.Stats mock due to complex union type requirements
// Mock file system operations
vitest_1.vi.mock('fs/promises', () => ({
    readFile: vitest_1.vi.fn(),
    writeFile: vitest_1.vi.fn(),
    mkdir: vitest_1.vi.fn(),
    stat: vitest_1.vi.fn(),
}));
vitest_1.vi.mock('path', () => ({
    join: vitest_1.vi.fn(),
    dirname: vitest_1.vi.fn(),
}));
(0, vitest_1.describe)('Evaluation Framework', () => {
    (0, vitest_1.describe)('GuardrailMetricsCalculator', () => {
        (0, vitest_1.it)('should calculate metrics correctly', () => {
            const calculator = new evals_1.GuardrailMetricsCalculator();
            const results = [
                {
                    id: '1',
                    expectedTriggers: { test: true },
                    triggered: { test: true },
                    details: {},
                },
                {
                    id: '2',
                    expectedTriggers: { test: false },
                    triggered: { test: false },
                    details: {},
                },
                {
                    id: '3',
                    expectedTriggers: { test: false },
                    triggered: { test: true },
                    details: {},
                },
                {
                    id: '4',
                    expectedTriggers: { test: true },
                    triggered: { test: false },
                    details: {},
                },
            ];
            const metrics = calculator.calculate(results);
            (0, vitest_1.expect)(metrics).toHaveProperty('test');
            const testMetrics = metrics['test'];
            (0, vitest_1.expect)(testMetrics.truePositives).toBe(1);
            (0, vitest_1.expect)(testMetrics.falsePositives).toBe(1);
            (0, vitest_1.expect)(testMetrics.falseNegatives).toBe(1);
            (0, vitest_1.expect)(testMetrics.trueNegatives).toBe(1);
            (0, vitest_1.expect)(testMetrics.precision).toBe(0.5);
            (0, vitest_1.expect)(testMetrics.recall).toBe(0.5);
            (0, vitest_1.expect)(testMetrics.f1Score).toBe(0.5);
        });
        (0, vitest_1.it)('should handle empty results', () => {
            const calculator = new evals_1.GuardrailMetricsCalculator();
            (0, vitest_1.expect)(() => calculator.calculate([])).toThrow('Cannot calculate metrics for empty results list');
        });
        (0, vitest_1.it)('should handle single result', () => {
            const calculator = new evals_1.GuardrailMetricsCalculator();
            const results = [
                {
                    id: '1',
                    expectedTriggers: { test: true },
                    triggered: { test: true },
                    details: {},
                },
            ];
            const metrics = calculator.calculate(results);
            const testMetrics = metrics['test'];
            (0, vitest_1.expect)(testMetrics.truePositives).toBe(1);
            (0, vitest_1.expect)(testMetrics.falsePositives).toBe(0);
            (0, vitest_1.expect)(testMetrics.falseNegatives).toBe(0);
            (0, vitest_1.expect)(testMetrics.trueNegatives).toBe(0);
        });
    });
    (0, vitest_1.describe)('validateDataset', () => {
        (0, vitest_1.it)('should validate valid dataset', async () => {
            const mockFs = await Promise.resolve().then(() => __importStar(require('fs/promises')));
            vitest_1.vi.mocked(mockFs.stat).mockResolvedValue({
                isFile: () => true,
                isDirectory: () => false,
                size: 1024
            });
            vitest_1.vi.mocked(mockFs.readFile).mockResolvedValue('{"id":"1","data":"Sample 1","expectedTriggers":{"test":true}}\n{"id":"2","data":"Sample 2","expectedTriggers":{"test":false}}');
            const [isValid, errors] = await (0, evals_1.validateDataset)('/tmp/test.jsonl');
            (0, vitest_1.expect)(isValid).toBe(true);
            (0, vitest_1.expect)(errors).toHaveLength(1);
            (0, vitest_1.expect)(errors[0]).toBe('Validation successful!');
        });
        (0, vitest_1.it)('should validate dataset with snake_case field names', async () => {
            const mockFs = await Promise.resolve().then(() => __importStar(require('fs/promises')));
            vitest_1.vi.mocked(mockFs.stat).mockResolvedValue({
                isFile: () => true,
                isDirectory: () => false,
                size: 1024
            });
            vitest_1.vi.mocked(mockFs.readFile).mockResolvedValue('{"id":"1","data":"Sample 1","expected_triggers":{"test":true}}\n{"id":"2","data":"Sample 2","expected_triggers":{"test":false}}');
            const [isValid, errors] = await (0, evals_1.validateDataset)('/tmp/test.jsonl');
            (0, vitest_1.expect)(isValid).toBe(true);
            (0, vitest_1.expect)(errors).toHaveLength(1);
            (0, vitest_1.expect)(errors[0]).toBe('Validation successful!');
        });
        (0, vitest_1.it)('should validate dataset with mixed field naming conventions', async () => {
            const mockFs = await Promise.resolve().then(() => __importStar(require('fs/promises')));
            vitest_1.vi.mocked(mockFs.stat).mockResolvedValue({
                isFile: () => true,
                isDirectory: () => false,
                size: 1024
            });
            vitest_1.vi.mocked(mockFs.readFile).mockResolvedValue('{"id":"1","data":"Sample 1","expectedTriggers":{"test":true}}\n{"id":"2","data":"Sample 2","expected_triggers":{"test":false}}');
            const [isValid, errors] = await (0, evals_1.validateDataset)('/tmp/test.jsonl');
            (0, vitest_1.expect)(isValid).toBe(true);
            (0, vitest_1.expect)(errors).toHaveLength(1);
            (0, vitest_1.expect)(errors[0]).toBe('Validation successful!');
        });
        (0, vitest_1.it)('should detect invalid dataset structure', async () => {
            const mockFs = await Promise.resolve().then(() => __importStar(require('fs/promises')));
            vitest_1.vi.mocked(mockFs.stat).mockResolvedValue({
                isFile: () => true,
                isDirectory: () => false,
                size: 1024
            });
            vitest_1.vi.mocked(mockFs.readFile).mockResolvedValue('{"id":"1","data":"Sample 1"}\n{"id":"2","expectedTriggers":{"test":false}}');
            const [isValid, errors] = await (0, evals_1.validateDataset)('/tmp/test.jsonl');
            (0, vitest_1.expect)(isValid).toBe(false);
            (0, vitest_1.expect)(errors.length).toBeGreaterThan(0);
        });
        (0, vitest_1.it)('should handle malformed JSON', async () => {
            const mockFs = await Promise.resolve().then(() => __importStar(require('fs/promises')));
            vitest_1.vi.mocked(mockFs.stat).mockResolvedValue({
                isFile: () => true,
                isDirectory: () => false,
                size: 1024
            });
            vitest_1.vi.mocked(mockFs.readFile).mockResolvedValue('invalid json\n{"id":"1","data":"Sample 1","expectedTriggers":{"test":true}}');
            const [isValid, errors] = await (0, evals_1.validateDataset)('/tmp/test.jsonl');
            (0, vitest_1.expect)(isValid).toBe(false);
            (0, vitest_1.expect)(errors.length).toBeGreaterThan(0);
        });
    });
    (0, vitest_1.describe)('JsonResultsReporter', () => {
        (0, vitest_1.it)('should save results to files', async () => {
            const mockFs = await Promise.resolve().then(() => __importStar(require('fs/promises')));
            const mockPath = await Promise.resolve().then(() => __importStar(require('path')));
            vitest_1.vi.mocked(mockFs.mkdir).mockResolvedValue(undefined);
            vitest_1.vi.mocked(mockFs.writeFile).mockResolvedValue(undefined);
            vitest_1.vi.mocked(mockPath.join).mockReturnValue('/tmp/results.jsonl');
            const reporter = new evals_1.JsonResultsReporter();
            const results = [
                {
                    id: '1',
                    expectedTriggers: { test: true },
                    triggered: { test: true },
                    details: {},
                },
            ];
            const metrics = {
                test: {
                    truePositives: 1,
                    falsePositives: 0,
                    falseNegatives: 0,
                    trueNegatives: 0,
                    totalSamples: 1,
                    precision: 1.0,
                    recall: 1.0,
                    f1Score: 1.0,
                },
            };
            await (0, vitest_1.expect)(reporter.save(results, metrics, 'test-output')).resolves.not.toThrow();
        });
        (0, vitest_1.it)('should create output directory if it does not exist', async () => {
            const mockFs = await Promise.resolve().then(() => __importStar(require('fs/promises')));
            const mockPath = await Promise.resolve().then(() => __importStar(require('path')));
            vitest_1.vi.mocked(mockFs.mkdir).mockResolvedValue(undefined);
            vitest_1.vi.mocked(mockFs.writeFile).mockResolvedValue(undefined);
            vitest_1.vi.mocked(mockPath.join).mockReturnValue('/tmp/results.jsonl');
            const reporter = new evals_1.JsonResultsReporter();
            const results = [
                {
                    id: '1',
                    expectedTriggers: { test: true },
                    triggered: { test: true },
                    details: {},
                },
            ];
            const metrics = {
                test: {
                    truePositives: 1,
                    falsePositives: 0,
                    falseNegatives: 0,
                    trueNegatives: 0,
                    totalSamples: 1,
                    precision: 1.0,
                    recall: 1.0,
                    f1Score: 1.0,
                },
            };
            await (0, vitest_1.expect)(reporter.save(results, metrics, 'new-output-dir')).resolves.not.toThrow();
        });
        (0, vitest_1.it)('should reject empty results', async () => {
            const reporter = new evals_1.JsonResultsReporter();
            const results = [];
            const metrics = {};
            await (0, vitest_1.expect)(reporter.save(results, metrics, 'test-output')).rejects.toThrow('Cannot save empty results list');
        });
    });
});
