// @openai/guardrails/dist/__tests__/unit/registry.test.js
"use strict";
/**
 * Unit tests for the registry module.
 *
 * This module tests the guardrail registry functionality including:
 * - Registration and retrieval
 * - Enumeration and metadata
 * - Configuration schemas
 * - Context requirements
 * - Overwriting behavior
 */
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
const zod_1 = require("zod");
const registry_1 = require("../../registry");
const spec_1 = require("../../spec");
// Zod schemas for testing
const TestConfigSchema = zod_1.z.object({
    threshold: zod_1.z.number(),
    enabled: zod_1.z.boolean(),
});
const TestContextSchema = zod_1.z.object({
    testProperty: zod_1.z.string(),
});
// Removed unused schemas
// Mock check function for testing
const mockCheck = vitest_1.vi.fn().mockImplementation(() => ({
    tripwireTriggered: false,
}));
(0, vitest_1.describe)('Registry Module', () => {
    let registry;
    (0, vitest_1.beforeEach)(() => {
        registry = new registry_1.GuardrailRegistry();
        vitest_1.vi.clearAllMocks();
    });
    (0, vitest_1.describe)('GuardrailRegistry', () => {
        (0, vitest_1.it)('should register and retrieve guardrails', () => {
            registry.register('test_guard', mockCheck, 'Test guardrail', 'text/plain');
            const spec = registry.get('test_guard');
            (0, vitest_1.expect)(spec).toBeDefined();
            (0, vitest_1.expect)(spec?.name).toBe('test_guard');
            (0, vitest_1.expect)(spec?.description).toBe('Test guardrail');
        });
        (0, vitest_1.it)('should return undefined for non-existent guardrails', () => {
            const spec = registry.get('non_existent');
            (0, vitest_1.expect)(spec).toBeUndefined();
        });
        (0, vitest_1.it)('should enumerate all registered guardrails', () => {
            registry.register('guard1', mockCheck, 'First guard', 'text/plain');
            registry.register('guard2', mockCheck, 'Second guard', 'text/plain');
            const specs = registry.all();
            (0, vitest_1.expect)(specs).toHaveLength(2);
            (0, vitest_1.expect)(specs.map((s) => s.name)).toContain('guard1');
            (0, vitest_1.expect)(specs.map((s) => s.name)).toContain('guard2');
        });
        (0, vitest_1.it)('should handle guardrail with metadata', () => {
            const metadata = {
                engine: 'typescript',
                version: '1.0.0',
            };
            registry.register('metadata_guard', mockCheck, 'Guard with metadata', 'text/plain', undefined, undefined, metadata);
            const spec = registry.get('metadata_guard');
            (0, vitest_1.expect)(spec?.metadata?.engine).toBe('typescript');
            (0, vitest_1.expect)(spec?.metadata?.version).toBe('1.0.0');
        });
        (0, vitest_1.it)('should handle guardrail with config schema', () => {
            registry.register('schema_guard', mockCheck, 'Guard with schema', 'text/plain', TestConfigSchema);
            const spec = registry.get('schema_guard');
            (0, vitest_1.expect)(spec?.configSchema).toBe(TestConfigSchema);
        });
        (0, vitest_1.it)('should handle guardrail with context requirements', () => {
            registry.register('context_guard', mockCheck, 'Guard with context', 'text/plain', undefined, TestContextSchema);
            const spec = registry.get('context_guard');
            (0, vitest_1.expect)(spec?.ctxRequirements).toBe(TestContextSchema);
        });
        (0, vitest_1.it)('should allow overwriting existing guardrails', () => {
            registry.register('overwrite_test', mockCheck, 'First version', 'text/plain');
            registry.register('overwrite_test', mockCheck, 'Second version', 'text/plain');
            const spec = registry.get('overwrite_test');
            (0, vitest_1.expect)(spec?.description).toBe('Second version');
        });
        (0, vitest_1.it)('should handle empty registry', () => {
            const specs = registry.all();
            (0, vitest_1.expect)(specs).toHaveLength(0);
        });
        (0, vitest_1.it)('should handle registry with single guardrail', () => {
            registry.register('single_guard', mockCheck, 'Single guard', 'text/plain');
            const specs = registry.all();
            (0, vitest_1.expect)(specs).toHaveLength(1);
            (0, vitest_1.expect)(specs[0].name).toBe('single_guard');
        });
    });
    (0, vitest_1.describe)('GuardrailSpec', () => {
        (0, vitest_1.it)('should create spec with all properties', () => {
            const metadata = {
                engine: 'typescript',
            };
            const spec = new spec_1.GuardrailSpec('full_spec', 'Full specification', 'text/plain', TestConfigSchema, mockCheck, TestContextSchema, metadata);
            (0, vitest_1.expect)(spec.name).toBe('full_spec');
            (0, vitest_1.expect)(spec.description).toBe('Full specification');
            (0, vitest_1.expect)(spec.mediaType).toBe('text/plain');
            (0, vitest_1.expect)(spec.checkFn).toBe(mockCheck);
            (0, vitest_1.expect)(spec.metadata).toBe(metadata);
        });
        (0, vitest_1.it)('should instantiate guardrail from spec', () => {
            const spec = new spec_1.GuardrailSpec('test_spec', 'Test specification', 'text/plain', TestConfigSchema, mockCheck, TestContextSchema);
            const guardrail = spec.instantiate({ threshold: 5, enabled: true });
            (0, vitest_1.expect)(guardrail.definition).toBe(spec);
            (0, vitest_1.expect)(guardrail.config).toEqual({ threshold: 5, enabled: true });
        });
        (0, vitest_1.it)('should run instantiated guardrail', async () => {
            const spec = new spec_1.GuardrailSpec('test_spec', 'Test specification', 'text/plain', TestConfigSchema, mockCheck, TestContextSchema);
            const guardrail = spec.instantiate({ threshold: 5, enabled: true });
            const result = await guardrail.run({ testProperty: 'test' }, 'Hello world');
            (0, vitest_1.expect)(result.tripwireTriggered).toBe(false);
            (0, vitest_1.expect)(mockCheck).toHaveBeenCalledWith({ testProperty: 'test' }, 'Hello world', { threshold: 5, enabled: true });
        });
    });
});
