// @openai/guardrails/dist/__tests__/unit/types.test.js
"use strict";
/**
 * Unit tests for the types module.
 *
 * This module tests the core type definitions including:
 * - GuardrailResult structure and immutability
 * - CheckFn function signatures
 * - Context interfaces
 * - Type compatibility
 */
Object.defineProperty(exports, "__esModule", { value: true });
const vitest_1 = require("vitest");
(0, vitest_1.describe)('Types Module', () => {
    (0, vitest_1.describe)('GuardrailResult', () => {
        (0, vitest_1.it)('should create result with required fields', () => {
            const result = {
                tripwireTriggered: true,
                info: {
                    guardrail_name: 'Example',
                },
            };
            (0, vitest_1.expect)(result.tripwireTriggered).toBe(true);
            (0, vitest_1.expect)(result.info.guardrail_name).toBe('Example');
        });
        (0, vitest_1.it)('should create result with custom info', () => {
            const info = { reason: 'test', severity: 'high' };
            const result = {
                tripwireTriggered: false,
                info: {
                    guardrail_name: 'Example',
                    checked_text: 'test',
                    ...info,
                },
            };
            (0, vitest_1.expect)(result.tripwireTriggered).toBe(false);
            (0, vitest_1.expect)(result.info.reason).toBe('test');
            (0, vitest_1.expect)(result.info.severity).toBe('high');
            (0, vitest_1.expect)(result.info.checked_text).toBe('test');
        });
        (0, vitest_1.it)('should handle minimal info', () => {
            const result = {
                tripwireTriggered: true,
                info: {
                    guardrail_name: 'Example',
                },
            };
            (0, vitest_1.expect)(result.tripwireTriggered).toBe(true);
            (0, vitest_1.expect)(result.info.guardrail_name).toBe('Example');
        });
    });
    (0, vitest_1.describe)('CheckFn', () => {
        (0, vitest_1.it)('should work with sync function', () => {
            const syncCheck = (ctx, data) => ({
                tripwireTriggered: data === 'trigger',
                info: {
                    guardrail_name: 'Sync',
                },
            });
            const result = syncCheck({}, 'trigger');
            (0, vitest_1.expect)(result.tripwireTriggered).toBe(true);
        });
        (0, vitest_1.it)('should work with async function', async () => {
            const asyncCheck = async (ctx, data) => ({
                tripwireTriggered: data === 'trigger',
                info: {
                    guardrail_name: 'Async',
                },
            });
            const result = await asyncCheck({}, 'trigger');
            (0, vitest_1.expect)(result.tripwireTriggered).toBe(true);
        });
    });
    (0, vitest_1.describe)('GuardrailLLMContext', () => {
        (0, vitest_1.it)('should require guardrailLlm property', () => {
            const context = {
                guardrailLlm: {},
            };
            (0, vitest_1.expect)(context.guardrailLlm).toBeDefined();
        });
        (0, vitest_1.it)('should work with mock LLM client', () => {
            // Test that the interface can be implemented with any object that has guardrailLlm
            const mockLLM = { someMethod: () => 'test' };
            const context = {
                guardrailLlm: mockLLM,
            };
            (0, vitest_1.expect)(context.guardrailLlm).toBeDefined();
            (0, vitest_1.expect)(context.guardrailLlm.someMethod()).toBe('test');
        });
    });
    (0, vitest_1.describe)('Type compatibility', () => {
        (0, vitest_1.it)('should allow flexible context types', () => {
            const check = (ctx, data, config) => ({
                tripwireTriggered: data.length > config.threshold,
                info: {
                    guardrail_name: 'Length',
                },
            });
            const result = check({ user: 'test' }, 'hello', { threshold: 3 });
            (0, vitest_1.expect)(result.tripwireTriggered).toBe(true);
        });
        (0, vitest_1.it)('should allow flexible input types', () => {
            const check = (ctx, data, _config) => ({
                tripwireTriggered: false,
                info: {
                    guardrail_name: 'FlexibleInput',
                    preview: String(data),
                },
            });
            const result = check({}, 'string input', {});
            (0, vitest_1.expect)(result.tripwireTriggered).toBe(false);
        });
        (0, vitest_1.it)('should allow flexible config types', () => {
            const check = (ctx, data, _config) => ({
                tripwireTriggered: false,
                info: {
                    guardrail_name: 'FlexibleConfig',
                    preview: String(data),
                },
            });
            const result = check({}, 'input', { complex: { nested: 'config' } });
            (0, vitest_1.expect)(result.tripwireTriggered).toBe(false);
        });
    });
});
